File: 프로그래머스/5/49190. 방의 개수/방의 개수.py
import sys
sys.setrecursionlimit(10 ** 6)

class DisjointSet:
    def __init__(self, nodes):
        self.parent = { node: node for node in nodes }
        
    def find(self, a):
        if a != self.parent[a]:
            self.parent[a] = self.find(self.parent[a])
            return self.parent[a]
        return a
    
    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)
        
        if root_a == root_b:
            return
        
        self.parent[root_a] = root_b

def move(node, direction):
    x, y = node
    if direction == 0:
        return (x - 1, y), (x - 2, y)
    if direction == 1:
        return (x - 1, y + 1), (x - 2, y + 2)
    if direction == 2:
        return (x, y + 1), (x, y + 2)
    if direction == 3:
        return (x + 1, y + 1), (x + 2, y + 2)
    if direction == 4:
        return (x + 1, y), (x + 2, y)
    if direction == 5:
        return (x + 1, y - 1), (x + 2, y - 2)
    if direction == 6:
        return (x, y - 1), (x, y - 2)
    return (x - 1, y - 1), (x - 2, y - 2)
    

def solution(arrows):
    root = (0, 0)
    
    nodes = set()
    
    node = root
    nodes.add(root)
    
    
    for direction in arrows:
        next_node, next_next_node = move(node, direction)
        nodes.add(next_node)
        nodes.add(next_next_node)
        node = next_next_node
        
        
    disjointSet = DisjointSet(nodes)    
    
    visited_edges = set()
    
    node = root
    count = 0
    
    
    for direction in arrows:
        next_node, next_next_node = move(node, direction)
        
        edge = tuple(sorted((node, next_node)))
        if edge not in visited_edges and disjointSet.find(node) == disjointSet.find(next_node):
            count += 1
            
        visited_edges.add(edge)
        disjointSet.union(node, next_node)
        
        node = next_node
        
        edge = tuple(sorted((next_node, next_next_node)))
        if edge not in visited_edges and disjointSet.find(next_node) == disjointSet.find(next_next_node):
            count += 1
            
        visited_edges.add(edge)
        disjointSet.union(next_node, next_next_node)
        
        node = next_next_node

    return count
        
        
        


    
    
    
    

# 그래프에서 사이클의 개수를 찾는 문제
# disjoint set에서, find(current) == find(next)인 경우
# - 사이클의 개수는 방은 항상 한번에 한개씩만 추가됌
# - 예외사항: 1 => 2 => 1 같은 것은 사이클은 되나 방이 안 만들어짐
#   처리 방법: edge의 방문 여부도 따로 확인
#            sort를 통해 vertex의 순서 보장, == 연산 작동 여부 확인함

File: 프로그래머스/5/49190. 방의 개수/방의 개수.kt
class Solution {
    fun move(current: Pair<Int, Int>, direction: Int): List<Pair<Int, Int>> {
        val (x, y) = current
        val next = when (direction) {
            0 -> listOf(x - 1 to y, x - 2 to y)
            1 -> listOf(x - 1 to y + 1, x - 2 to y + 2)
            2 -> listOf(x to y + 1, x to y + 2)
            3 -> listOf(x + 1 to y + 1, x + 2 to y + 2)
            4 -> listOf(x + 1 to y, x + 2 to y)
            5 -> listOf(x + 1 to y - 1, x + 2 to y - 2)
            6 -> listOf(x to y - 1, x to y - 2)
            else -> listOf(x - 1 to y - 1, x - 2 to y - 2)
        }
        return next
    }
    
    fun solution(arrows: IntArray): Int {
        val start = 0 to 0
        
        val path = arrows.fold(mutableListOf<Pair<Int, Int>>(start)) { acc, arrow ->
            val current = acc[acc.size - 1]
            val next_nodes = move(current, arrow)
            next_nodes.forEach { acc.add(it) }
            acc
        }
        
        val nodes = path.toSet()
        
        val disjointSet = DisjointSet(nodes)
        val visitedEdges = mutableSetOf<List<Pair<Int, Int>>>()
    
        var current = 0 to 0
        var count = 0
        
        val comparator = Comparator<Pair<Int, Int>> { a, b ->
            if (a.first != b.first) {
                a.first - b.first
            } else {
                a.second - b.second
            }
        }
        
        for (node in path) {
            if (node == current) {
                continue
            }
            
            val edge = listOf(current, node).sortedWith(comparator)
            
            if (edge !in visitedEdges && disjointSet.find(current) == disjointSet.find(node)) {
                count += 1
            }
            
            visitedEdges.add(edge)
            disjointSet.union(current, node)
            current = node
        }
        
        
        return count
    }
}

class DisjointSet(nodes: Set<Pair<Int, Int>>) {
    val parent = nodes.fold(mutableMapOf<Pair<Int, Int>, Pair<Int,Int>>()) { acc, node ->
        acc[node] = node
        acc
    }
    
    fun find(node: Pair<Int, Int>): Pair<Int, Int> {
        if (node == parent[node]!!) {
            return node
        } else {
            parent[node] = find(parent[node]!!)
            return parent[node]!!
        }
    }
    
    fun union(node1: Pair<Int, Int>, node2: Pair<Int, Int>) {
        val root1 = find(node1)
        val root2 = find(node2)
        
        if (root1 == root2) {
            return
        }
        parent[root1] = root2
    }
}


// arrows로 path 안에 있는 node들 모두 찾기
// cycle이 만들어질 때마다 += 1
// cycle 판별은 Disjoint Set으로 수행